onnxruntime/core/platform/windows/env.cc

```diff
@@ -161,21 +156,50 @@ class WindowsThread : public EnvThread {
     }
     unsigned ret = 0;
     ORT_TRY {
-      // TODO: should I try to use SetThreadSelectedCpuSets?
       if (p->affinity.has_value() && !p->affinity->empty()) {
-        DWORD_PTR mask = 0;
-        for (auto id : *p->affinity) {
-          mask |= DWORD_PTR{1} << id;
-        }
-        auto rc = SetThreadAffinityMask(GetCurrentThread(), mask);
-        if (!rc) {
-          const auto error_code = GetLastError();
-          LOGS_DEFAULT(ERROR) << "SetThreadAffinityMask failed for thread: " << GetCurrentThreadId()
-                              << ", index: " << p->index
-                              << ", mask: " << *p->affinity
-                              << ", error code: " << error_code
-                              << ", error msg: " << std::system_category().message(error_code)
-                              << ". Specify the number of threads explicitly so the affinity is not set.";
+        int group_id = -1;
+        KAFFINITY mask = 0;
+        KAFFINITY bit = 1;
+        const WindowsEnv& env = WindowsEnv::Instance();
+        for (auto global_processor_id : *p->affinity) {
+          auto processor_info = env.GetProcessorAffinityMask(global_processor_id);
+          if (processor_info.local_processor_id > -1) {
+            mask |= bit << processor_info.local_processor_id;
+          } else {
+            LOGS_DEFAULT(ERROR) << "Cannot set affinity for thread " << GetCurrentThreadId()
+                                << ", processor " << global_processor_id << " not found";
+            group_id = -1;
+            mask = 0;
+            break;
+          }
+          if (group_id == -1) {
+            group_id = processor_info.group_id;
+          } else if (group_id != processor_info.group_id) {
+            LOGS_DEFAULT(ERROR) << "Cannot set cross-group affinity for thread "
+                                << GetCurrentThreadId();
+            group_id = -1;
+            mask = 0;
+            break;
+          }
+        }  //for
+        if (group_id > -1 && mask) {
+          GROUP_AFFINITY thread_affinity;
+          memset(&thread_affinity, 0x0, sizeof(GROUP_AFFINITY));
+          thread_affinity.Group = static_cast<WORD>(group_id);
```

Comment:
>group_id [](http://example.com/codeflow?start=52&length=8)

group_id is int. I think this should be checked against the max() of WORD. We are checking local_processor IDs, but we are not checking group ID which may come from miscalculated customer input.

Reply:
group id is not a customer input, we manage that totally.

Reply:
I do not understand. How come then, group_id can be mismatched on line 179?

Reply:
So how is 1 and 2 possible? Would this be a bug when generating the map?

Reply:
1 and 2 is possible because the "global_processor_id" for each thread are directly read from the affinity string set by the user, which might contain the wrong configuration.